abci/auction.go

```diff
@@ -0,0 +1,238 @@
+	// bundled transactions are valid.
+	var topOfBlock TopOfBlock
+	for _, bidTx := range bidTxs {
+		// Cache the context so that we can write it back to the original context
+		// when we know we have a valid top of block bundle.
+		cacheCtx, write := ctx.CacheContext()
+
+		// Attempt to build the top of block using the bid transaction.
+		proposal, err := h.buildTOB(cacheCtx, bidTx)
+		if err != nil {
+			h.logger.Info(
+				"vote extension auction failed to verify auction tx",
+				"err", err,
+			)
+			txsToRemove[bidTx] = struct{}{}
+			continue
+		}
+
+		if proposal.Size <= maxBytes {
+			// At this point, both the bid transaction itself and all the bundled
+			// transactions are valid. So we select the bid transaction along with
+			// all the bundled transactions and apply the state changes to the cache
+			// context.
+			topOfBlock = proposal
+			write()
+
+			break
+		}
+
+		h.logger.Info(
+			"failed to select auction bid tx; auction tx size is too large",
+			"tx_size", proposal.Size,
+			"max_size", maxBytes,
+		)
+	}
+
+	// Remove all of the transactions that were not valid.
+	for tx := range txsToRemove {
+		h.RemoveTx(tx)
+	}
+
+	return topOfBlock
+}
+
+// VerifyTOB verifies that the set of vote extensions used in prepare proposal deterministically
+// produce the same top of block proposal.
+func (h *ProposalHandler) VerifyTOB(ctx sdk.Context, proposal [][]byte) (*AuctionInfo, error) {
+	// Proposal must include at least the auction info.
+	if len(proposal) < MinProposalSize {
+		return nil, fmt.Errorf("proposal is too small; expected at least %d slots", MinProposalSize)
+	}
+
+	// Extract the auction info from the proposal.
+	auctionInfo := AuctionInfo{}
+	if err := auctionInfo.Unmarshal(proposal[AuctionInfoIndex]); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal auction info: %w", err)
+	}
+
+	// Verify that the proposal contains the expected number of top of block transactions.
+	if len(proposal) < int(auctionInfo.NumTxs)+MinProposalSize {
+		return nil, fmt.Errorf("number of txs in proposal do not match expected in auction info; expected at least %d slots", auctionInfo.NumTxs+MinProposalSize)
+	}
+
+	// Unmarshall the vote extension information from the auction info.
+	lastCommitInfo := abci.ExtendedCommitInfo{}
+	if err := lastCommitInfo.Unmarshal(auctionInfo.ExtendedCommitInfo); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal last commit info from auction info: %w", err)
+	}
+
+	// Build the top of block proposal from the auction info.
+	expectedTOB := h.BuildTOB(ctx, lastCommitInfo, auctionInfo.MaxTxBytes)
+
+	// Verify that the top of block txs matches the top of block proposal txs.
+	actualTOBTxs := proposal[MinProposalSize : auctionInfo.NumTxs+MinProposalSize]
+	if !reflect.DeepEqual(actualTOBTxs, expectedTOB.Txs) {
+		return nil, fmt.Errorf("expected top of block txs does not match top of block proposal")
+	}
+
+	return &auctionInfo, nil
```

Comment:
Why do we need to return a pointer? Maybe just have the type creation be a pointer instead of taking its address here?